client/go/outline/vpn/vpn.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpn
+
+import (
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"sync"
+
+	"github.com/Jigsaw-Code/outline-sdk/network"
+)
+
+// Config holds the configuration to establish a system-wide [VPNConnection].
+type Config struct {
+	ID              string   `json:"id"`
+	InterfaceName   string   `json:"interfaceName"`
+	IPAddress       string   `json:"ipAddress"`
+	DNSServers      []string `json:"dnsServers"`
+	ConnectionName  string   `json:"connectionName"`
+	RoutingTableId  uint32   `json:"routingTableId"`
+	RoutingPriority uint32   `json:"routingPriority"`
+	ProtectionMark  uint32   `json:"protectionMark"`
+}
+
+// Status defines the possible states of a [VPNConnection].
+type Status string
+
+// Constants representing the different VPN connection statuses.
+const (
+	StatusUnknown       Status = "Unknown"
+	StatusConnected     Status = "Connected"
+	StatusDisconnected  Status = "Disconnected"
+	StatusConnecting    Status = "Connecting"
+	StatusDisconnecting Status = "Disconnecting"
+)
+
+// ProxyDevice is an interface representing a remote proxy server device.
+type ProxyDevice interface {
+	network.IPDevice
+
+	// Connect establishes a connection to the proxy device.
+	Connect(ctx context.Context) error
+
+	// SupportsUDP returns true if the proxy device is able to handle UDP traffic.
+	SupportsUDP() bool
+
+	// RefreshConnectivity refreshes the UDP support of the proxy device.
+	RefreshConnectivity(ctx context.Context) error
+}
+
+// platformVPNConn is an interface representing an OS-specific VPN connection.
+type platformVPNConn interface {
+	// Establish creates a TUN device and routes all system traffic to it.
+	Establish(ctx context.Context) error
+
+	// TUN returns a L3 IP tun device associated with the VPN connection.
+	TUN() io.ReadWriteCloser
+
+	// Close terminates the VPN connection and closes the TUN device.
+	Close() error
+}
+
+// VPNConnection represents a system-wide VPN connection.
+type VPNConnection struct {
+	ID          string `json:"id"`
+	Status      Status `json:"status"`
+	SupportsUDP *bool  `json:"supportsUDP"`
```

Comment:
It doesn't seem to be used. Delete?

We seem to use it when we return the json on connect, but that's the only place. We can get it from the Remote Device instead.

Reply:
We are not, this is for potential UI display purpose, we can remove it for now.

Reply:
Deleted.